### ISSUE TITLE
Rotation transform for Rectangle Editor

### ISSUE THREAD

**Issue by richardkc** [annotorious/annotorious#344]
There's any way to rotate rect? such like transform. I cannot found any way to do that, unless customize polygon tool and then calculate the position of points.

**Comment by rsimon** [comment]
That's currently not available as a a feature, I'm afraid. But very likely to come further down the line in Annotorious v3.

**Comment by alexandrep-explorai** [comment]
@rsimon I know you probably have loads of other priorities, but I was wondering if this was still in the pipeline for future improvements to the library?

For a workaround I was thinking of just handling all rectangles in my app as polygons, and implementing a custom editor that could handle both editing rectangular polygons AND more complex polygons, as well as allow rotations.

**Comment by rsimon** [comment]
Yep, the idea would be to have the usual rotation handle on the rectangle (possibly snapping to 45° degrees on Shift). I wouldn't  store it as a polygon shape though - that would just muddy the model I think. Instead I'd just add a rotation field to the rectangle shape type, I guess.

In terms of implementation: not sure when I'd get to it. But TBH probably not until a sponsoring client comes along who funds the time.

**Comment by alexandrep-explorai** [comment]
Oh yea, it would be way better to have the information within the rectangle shape itself, I agree. I was just thinking of ways to achieve this within my own project without needing to drastically modify the shape models and the rendering in the core of the library.

**Comment by dragonchess63** [comment]
@rsimon How much would you take for the implementation and in what timeframe would it be ?

**Comment by rsimon** [comment]
Hi @dragonchess63, if you're interested in sponsoring, feel free to drop me a line at hello@rainersimon.io.

**Comment by rsimon** [comment]
Added in Annotorious v3.8.0 🚀

**Comment by tgordienko** [comment]
Hi, @rsimon How can I disable this feature?

**Comment by rsimon** [comment]
I'm afraid it's part of the new default rectangle editor (and thus similar to most other shape editing tools out there.) I guess we _could_ add a config option to toggle it later, if there's demand. But for now: if you really want to get rid of it, the only workaround would be to downgrade to the latest Annotorious 3.7 version ([3.7.22](https://www.npmjs.com/package/@annotorious/react/v/3.7.22)). No difference to 3.8 except for the rotation feature.

**Comment by MatthiasDurivet** [comment]
Hi @rsimon amazing to see this feature added, you have no idea how imaculate the timing is :)

I am having some trouble getting rotations to work in the Svelte + OSD context. I do not see the rotate handle appearing after I upgraded to v3.8. I am able to manually set the `rot` field and see the effect on the canvas.
I'm also unable to find any documentation on how exactly I'm supposed to use rotations, I suppose it should just give a rotation handle by default and I don't need to worry about anything else?

Is it a known limitation that the rotation handle doesn't show up in the Svelte version?

Thanks again for the continuous effort on this project!

**Comment by rsimon** [comment]
The rotation handle is "just there" from now on. It's not separate from the rest of the shape editor (corner and edge handles). Are you working off a local build? In this case, are you sure you built your `core` and `annotorious` packages?

The Svelte package is effectively unmaintained. But I don't think it would be possible to get the "old" rectangle editor without the new rotation handle – unless you're somehow importing an outdated (v3.7.x) version of `@annotorious/annotorious`.

As for your question: yes, docs still need updating. But, the gist: `rot` is now an (optional) field on the rectangle geometry, and you should be able to pass in any value in radians. The tool itself, however, will always give a (rad) rotation between 0 and 2 * PI.

**Comment by alexandrep-explorai** [comment]
> As for your question: yes, docs still need updating. But, the gist: `rot` is now an (optional) field on the rectangle geometry, and you should be able to pass in any value in radians. The tool itself, however, will always give a (rad) rotation between 0 and 2 * PI.

@rsimon Just to be a bit more precise, I'm pretty sure while testing the new feature I noted that rotations between 270 (exclusively) and 0 degrees were represented by values from -PI/2 to 0 rad. Not a bother, but I'm now wondering if this is intended since your comment here seems to indicate otherwise. The real values are from -PI/2 (-90 degrees) to 3PI/2 (270 degrees).

**Comment by rsimon** [comment]
Yikes - indeed. Hm, as you say: not a huge issue since it won't break anything. But I'll look into that. Normalizing to 0/2PI seems more sensible. Must have messed up my sin/cos somewhere.

**Comment by rsimon** [comment]
Looked into this, and indeed the range was theoretically [-2PI, 2PI], and [-PI/2, 3PI/2] in practice which (I believe) was related to the rotation happening around a fixed point. Anyways: the `rot` value is now properly normalized to `[0, 2PI)`.

This shouldn't break any existing functionality, and will be available along with the next release.

**Comment by alexandrep-explorai** [comment]
Sweet! Thanks for the update.
